accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,144 @@
+
+```c#
+    public readonly partial struct ActivityContext : IEquatable<ActivityContext>
+    {
+        // This constructor already exist but we are adding extra defaulted parameter for isRemote.
+        public ActivityContext(ActivityTraceId traceId, ActivitySpanId spanId, ActivityTraceFlags traceFlags, string? traceState = null, isRemote = false) {...}
+
+        // This is the new property
+        public bool IsRemote { get; }
+    }
+```
+
+## Activity.Kind Setter
+
+Activity.Kind is already exposed property but we have exposed it as a read-only property. We got OT adapters scenario that listen to the old created activities (by asp.net and Http clients). As the old Activities didn't support Kind property, the adapter needs to set this property before sending the activity to the sampling in the OT SDK. The proposal here is to expose the setter of this property.
+
+```c#
+    public partial class Activity : IDisposable
+    {
+        // The setter was private and now we are making it public
+        public ActivityKind Kind { get; set; } = ActivityKind.Internal;
+    }
+```
+
+## ActivityAttributesCollection
+
+OT has the concept of [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-attributes) which is a list of key-value pairs mapping a string to some value. The value can be of different types (e.g. numeric, bool, string, arrays...etc.). 
+
+> ```json
+>      'x-forwarded-for': 'foo.bar'
+>      'timeout': 5
+> ```
+
+The attributes are used in multiple places in the OT (e.g. in Span, Event, and Link). Activity support Tags which similar concept but Tags is limited to string value types and not supporting any other types `IEnumerable<KeyValuePair<string, string?>> Tags`.  We'll need to extend the Tags concept in our APIs to support the OT scenarios. We'll do that by introducing the ActivityAttributesCollection and extending Activity class to support adding different types of attribute values.
+
+ActivityAttributesCollection is needed to ensure the OT specs behavior is implemented and avoid API users to use different type of collections (e.g. Dictionary<>, List<>...etc.) which not conforming with the OT specs. The OT specs require the attributes always stored in order and define behavior when adding values for keys which already existed in the list, it will override the old value. In addition it define the behavior what should happen when using null values.
+
+```c#
+    public class ActivityAttributesCollection : IReadOnlyCollection<KeyValuePair<string, object>>
+    {
+        public ActivityAttributesCollection() { }
+        public ActivityAttributesCollection(IEnumerable<KeyValuePair<string, object>>) { }
+
+        // If the key was not set before, it will get added to the collections with the input value.
+        // If the key was set before, the mapped value of this key will be updated with the new input value.
+        public void Set(string key, int value) { }
```

Comment:
Will we expose more overloads for bool, string etc for values?

Reply:
No. this intended to be object and not int. I'll fix that now.

Reply:
I fixed it. Thanks for the catch.